### PR TITLE
Allow jms/serializer v2.0 and willdurand/Hateoas v3.0

### PR DESCRIPTION
**Pull request by goetas**
This introduces support for  jms/serializer v2.0

It looks that because of dependencies, this will work only when both FOSRestBundle and BazingaHateoasBundle will support jms/serializer v2.0

Closes https://github.com/nelmio/NelmioApiDocBundle/issues/1464
Closes https://github.com/nelmio/NelmioApiDocBundle/issues/1437

### CODE REVIEW

File: ModelDescriber/JMSModelDescriber.php [hunk 1]
```diff
@@ -81,8 +81,9 @@ public function describe(Model $model, Schema $schema)
                 $previousGroups = $groups;
                 $groups = $groups[$item->name];
             } elseif (!isset($groups[$item->name]) && !empty($this->previousGroups[$model->getHash()])) {
-                // $groups = $this->previousGroups[spl_object_hash($model)]; use this for jms/serializer 2.0
-                $groups = false === $this->propertyTypeUsesGroups($item->type) ? null : [GroupsExclusionStrategy::DEFAULT_GROUP];
+                $groups = false === $this->propertyTypeUsesGroups($item->type)
```
Comment (calbrecht) [inline ModelDescriber/JMSModelDescriber.php]: This line is not covered by the testsuite.

**Comment by michalsingr** [comment]
Hello, any progress here? Support for the latest jms/serializer is really a need... Thanks

**Comment by michalsingr** [comment]
Hello, 2 months from opening this MR... Any progress? (guys, I'll send you some chocolate bar or name something else ;-) )

**Comment by goetas** [comment]
https://github.com/willdurand/BazingaHateoasBundle/pull/82 is blocking this PR. 

if you want to sponsor this PR, will be happy to continue working on it.

**Comment by goetas** [comment]
some progress has been made :)

**Comment by magnetik** [comment]
Is something needed ? any help wanted ?

**Comment by goetas** [comment]
Any chance to test Hateoas v3.0?

**Comment by magnetik** [comment]
Hum nope I'm not using it :/

**Comment by goetas** [comment]
have added tests for the new group strategies. however I'm not yet happy with the tests..
having all the model names JMSPicture2, JMSChatRoom3, JMSChatRoom3, JMSChatRoom2 makes a lot of confusion, and the names generation is not really predictable.. to test it properly I should use also custom names for the special groups..

File: Tests/Functional/JMSFunctionalTest.php [hunk 2]
```diff
@@ -46,7 +48,7 @@ public function testModeChatDocumentation()
             'type' => 'object',
             'properties' => [
                 'picture' => [
-                    '$ref' => '#/definitions/JMSPicture',
+                    '$ref' => '#/definitions/JMSPicture2',
```
Comment (goetas) [inline Tests/Functional/JMSFunctionalTest.php]: This change can be understood by looking directly at the commit https://github.com/nelmio/NelmioApiDocBundle/pull/1433/commits/fd7b5e167915c87473dff5a8d148abb7b6f111f6

We forgot the `$previosGroups` parameter when calling recursively the function.

**Comment by fabianoroberto** [comment]
Why this PR is not integrated in master?

**Comment by GuilhemN** [comment]
@fabianoroberto because I did not have the time so far to review it again. And you're not really saving me time by asking.
If you want the process to be faster please review this PR, or even just test it in your own projects and provide feedback.

I'll try to review this PR next week, thanks again @goetas :)

File: ModelDescriber/JMSModelDescriber.php [hunk 3]
```diff
@@ -81,22 +84,33 @@ public function describe(Model $model, Schema $schema)
...
             // read property options from Swagger Property annotation if it exists
-            if (null !== $item->reflection) {
-                $property = $properties->get($annotationsReader->getPropertyName($item->reflection, $name));
-                $annotationsReader->updateProperty($item->reflection, $property, $groups);
-            } else {
+            try {
+                if (true === $isJmsV1 && property_exists($item, 'reflection') && null !== $item->reflection) {
+                    $reflection = $item->reflection;
+                } elseif ($item instanceof VirtualProperty) {
+                    $reflection = new \ReflectionProperty($item->class, $item->name);
```
Comment (GuilhemN) [inline ModelDescriber/JMSModelDescriber.php]: This is not required, right? (Because there is the same line below)

**Comment by GuilhemN** [comment]
My last comment is minor so I'm merging your PR still, I'll submit a patch if you think it is right :) Thank you!

**Comment by fabianoroberto** [comment]
> My last comment is minor so I'm merging your PR still, I'll submit a patch if you think it is right :) Thank you!

Please create before a new tag 🤓

File: ModelDescriber/JMSModelDescriber.php (hunk 3 above)
Comment (goetas) [reply ModelDescriber/JMSModelDescriber.php]: true, my bad, i must have missed it

Comment (goetas) [reply ModelDescriber/JMSModelDescriber.php]: In reality I wanted to fix https://github.com/nelmio/NelmioApiDocBundle/issues/1412 but later i've realized that was not that easy (so i gave up but I forgot about it).

BTW, if you have a moment to give me your feedback on https://github.com/nelmio/NelmioApiDocBundle/issues/1412

Comment (GuilhemN) [reply ModelDescriber/JMSModelDescriber.php]: Fixed in https://github.com/nelmio/NelmioApiDocBundle/commit/0414277d0aba277102d68fd920f88e6e762637fe ;)

Comment (goetas) [reply ModelDescriber/JMSModelDescriber.php]: :+1:

**Comment by GuilhemN** [comment]
@goetas Is there any reason you added php 7.2 with low deps to travis ? It seems some dependencies are broken on it, see https://travis-ci.org/nelmio/NelmioApiDocBundle/jobs/492898414.

**Comment by goetas** [comment]
jms serializer & hateoas works only with php 7.2

so in this was we can test low deps both with jms v1 and v2

**Comment by GuilhemN** [comment]
Ok thanks, I guess we should just bump some dependencies (probably the phpunit bridge) to have them working with php 7.2.

I released 3.4 still : https://github.com/nelmio/NelmioApiDocBundle/releases/tag/v3.4.0

**Pull request merged**